src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,275 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Experimental.Collections
+{
+    public class DictionarySlim<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey>
+    {
+        const int DefaultPrimeSize = 11;
+        private int[] _buckets;
+        private Entry[] _entries;
+
+        private struct Entry
+        {
+            public TKey key;
+            public TValue value;
+            public int next;
+        }
+
+        public DictionarySlim()
+        {
+            _buckets = new int[DefaultPrimeSize];
+            _entries = new Entry[DefaultPrimeSize];
+        }
+
+        public DictionarySlim(int capacity)
+        {
+            capacity = HashHelpers.GetPrime(capacity);
+            _buckets = new int[capacity];
+            _entries = new Entry[capacity];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetBucketIndex(TKey key) => (key.GetHashCode() & 0x7FFFFFFF) % _buckets.Length;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetEntryIndex(TKey key) => _buckets[GetBucketIndex(key)] - 1;
+
+        public int Count { get; private set; }
+
+        public bool ContainsKey(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex(key);
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return true;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return false;
+        }
+
+        public TValue GetValueOrDefault(TKey key)
+        {
+            Entry[] entries = _entries;
+            int entryIndex = GetEntryIndex(key);
+
+            while (entryIndex != -1)
+            {
+                if (entries[entryIndex].key.Equals(key))
+                {
+                    return entries[entryIndex].value;
+                }
+                entryIndex = entries[entryIndex].next;
+            }
+
+            return default;
+        }
+
+        public ref TValue this[TKey key]
```

Comment:
Can you please add a  `public bool Remove(TKey key)` ?

Reply:
This will of course mean we will need to decide what to do with the deleted slots in `_entries` eg
1. expensively coalesce the Entries (so adds continue to go at the end) which is simple but slow
1. use a free list like Dictionary (and possibly use Resize to coalece: Dictionary does not)
1. some other scheme

I like the free list without coalesce, it is only one more test on add (since in the k-Nucleotide case there are no removes so no free list)

Reply:
I had a shot at implementing with a free list, which you are welcome to borrow: https://github.com/dotnet/corefxlab/commit/da5f15991f7a9778f65c948bd54a646adebc10a1

The enumerator has to know how to skip removes. Temporarily, I put the free list in a HashSet. A better approach might be to make `entry.next` 1-based, so that 0 could be used to distinguish unused (-1 would continue to mean _pointing to_ nothing). Then the enumerator could just go through `_entries` skipping everything that has `.next` of 0.

Reply:
It looks like there may be some impact on perf due to the one extra test in the indexer. A bit noisy though:
|         Method |     Size |        Mean |      Error |     StdDev |      Median | Scaled | ScaledSD |
|--------------- |--------- |------------:|-----------:|-----------:|------------:|-------:|---------:|
|          Hack9 |   250000 |    124.5 ms |   2.472 ms |   4.061 ms |    124.4 ms |   1.00 |     0.00 |
|     Dictionary |   250000 |    222.6 ms |   4.463 ms |  13.159 ms |    217.8 ms |   1.79 |     0.12 |
| DictionarySlim |   250000 |    129.2 ms |   2.565 ms |   3.839 ms |    129.3 ms |   1.04 |     0.05 |
|                |          |             |            |            |             |        |          |
|          Hack9 |  2500000 |    634.1 ms |  20.608 ms |  60.763 ms |    631.9 ms |   1.00 |     0.00 |
|     Dictionary |  2500000 |  1,423.7 ms |  39.513 ms | 116.504 ms |  1,450.1 ms |   2.27 |     0.28 |
| DictionarySlim |  2500000 |    731.1 ms |  25.931 ms |  76.459 ms |    717.9 ms |   1.16 |     0.16 |
|                |          |             |            |            |             |        |          |
|          Hack9 | 25000000 |  6,239.2 ms | 186.737 ms | 550.599 ms |  6,384.2 ms |   1.00 |     0.00 |
|     Dictionary | 25000000 | 14,451.4 ms | 275.006 ms | 337.732 ms | 14,508.7 ms |   2.34 |     0.23 |
| DictionarySlim | 25000000 |  6,982.1 ms | 149.035 ms | 432.377 ms |  7,054.4 ms |   1.13 |     0.13 |

I tried adding a bit of [by-ref in the indexer ](https://github.com/danmosemsft/corefxlab/commit/eb6e50b66518a63019b4d4a2e8caac4010c6325b) but it seemed like that may have made it worse:

|         Method |     Size |        Mean |      Error |     StdDev |      Median | Scaled | ScaledSD |
|--------------- |--------- |------------:|-----------:|-----------:|------------:|-------:|---------:|
| DictionarySlim | 25000000 |  7,154.3 ms | 141.673 ms | 401.903 ms |  7,228.1 ms |   1.17 |     0.12 |

Reply:
The freelist test could go inside the count test, making add cheaper but at the expense of memory locality in entries perhaps.